### PR TITLE
[WIP] Add Material `Badge` widget

### LINKED ISSUE #92257: Add badge widget
## Use case/Problem
I couldn't find a native way to add badges to icons without doing some workaround with extra widgets. This seem to be allowed by the material design guidelines and they even show examples using it. [Material design badging](https://material.io/develop/android/components/badging)

I also found some open issues related with this topic, such as [Add support for Badges on BottomNavigationBarItems](https://github.com/flutter/flutter/issues/31209).

## Proposal
Extend Icon() API adding a badge like the follow example:

```dart
Icon.badge(
	Icons.settings, 
	badge: Badge(
		visible: true,
		type: BadgeType.numeric,
		maxValue: 20
		value: 10),
);
```

Where the type would be enum with values like:

```dart
enum BadgeType {
	numeric,
	boolean,
}
```

#### In case of boolean 
the badge only would display the single point at the corner of the Icon as follows.

boolean_type

#### In case of numeric 
the badge would show the given value (or the max value with the + suffix if the value surpasses it)
numeric_type
numeric_type_max

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes https://github.com/flutter/flutter/issues/92257

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/navigation_bar.dart [hunk 1]
```diff
@@ -15,6 +15,33 @@ import 'navigation_bar_theme.dart';
...
+enum BadgePosition {
+  /// Badge is positioned at the top start of the [NavigationDestination].
+  topStart,
+  /// Badge is positioned at the top end of the [NavigationDestination].
+  topEnd,
+  /// Badge is positioned at the bottom start of the [NavigationDestination].
+  bottomStart,
+  /// Badge is positioned at the bottom end of the [NavigationDestination].
+  bottomEnd,
+}
```
Comment (HrX03) [inline packages/flutter/lib/src/material/navigation_bar.dart:43]: wouldn't something like AlignmentGeometry make more sense? would support rtl out of the box and be more flexible

Comment (TahaTesser) [reply packages/flutter/lib/src/material/navigation_bar.dart:43]: This follows the material badge alignment naming scheme. 
Also start and end naming convention is implemented for RTL support rather left or right

File: packages/flutter/lib/src/material/navigation_bar.dart [hunk 2]
```diff
@@ -230,6 +257,9 @@ class NavigationDestination extends StatelessWidget {
     this.selectedIcon,
     required this.label,
     this.tooltip,
+    this.showBadge = false,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/navigation_bar.dart:260]: This doesn't seem like a natural Flutter API. Why not give NavigationDestination a `Widget?` valued badge property and provide a standard NavigationBadge widget with the right default colors, shape etc?

Comment (TahaTesser) [reply packages/flutter/lib/src/material/navigation_bar.dart]: Thanks for the feedback. 

I've reworked Badge to be an independent widget and will be supported via `Widget? badge` parameter in `NavigationDestination` and `NavigationRailDestination` in separate PRs.

**Review by HansMuller — COMMENTED** [review]
This is getting better but it's still needs some refinement. The Badge widget shouldn't require a controller, just a child.  A factory like `Badge.count()`, that created a private Badge subclass might handle the special case where the Badge's child displays an integer with a limited number of digits.

Updating Badges based on application state, by rebuilding the badges themselves or some part of the navigation bar or its ancestors, should be up to the application.  An example that demonstrated a simple version of this would be helpful.

Since the typical use-case for a Badge is displaying a short string within a filled "stadium" shape, it makes sense for it to have shape, backgroundColor and textStyle parameters. The TextStyle would be used to wrap the child in a DefaultTextStyle. A TextStyle parameter is a little tricky because the common case is to _just_ override the default TextStyle's color. If you also provide a textColor parameter, then it should probably override the TextStyle's color, even if that's non-null.

This widget will need to deal with [textScaleFactor](https://api.flutter.dev/flutter/widgets/MediaQueryData/textScaleFactor.html) and (probably) [visualDensity](https://api.flutter.dev/flutter/material/ThemeData/visualDensity.html) to the extent that it constrains its layout to fit into a navigation bar.

**Review by HansMuller — COMMENTED** [review]
Badges might be applied to any widget, not just to a NavigationDestination. This widget would be more useful if it had a child and an alignment property. That way it could be used to decorate NavigationDestinations as well as any other widget.

**Comment by Hixie** [comment]
@TahaTesser Is this something you are still working on?

**Comment by TahaTesser** [comment]
> @TahaTesser Is this something you are still working on?

Thanks for the comment, yes, I'll update this PR by Monday

**Comment by TahaTesser** [comment]
Update:
Still working on this, currently trying to badge alignment relative child widget.

**Comment by TahaTesser** [comment]
I'm trying to figure out how to position the badge relative to a child in a cleaner way using render objects and make it customizable depending on the child's size so the badge can be used with any child widget.

My current solution is to place it inside a `Stack` with some padding but it won't flexible, and will require custom padding whenever the badge is used with a child bigger a normal such as in `NavigationRailDestination`.

**Comment by TahaTesser** [comment]
Closing this for now as this new feature is a low priority right now.

**Pull request closed without merging**